internal/gatewayapi/helpers.go

```diff
@@ -223,3 +223,26 @@ func hostnameMatchesWildcardHostname(hostname, wildcardHostname string) bool {
 	wildcardMatch := strings.TrimSuffix(hostname, strings.TrimPrefix(wildcardHostname, "*"))
 	return len(wildcardMatch) > 0
 }
+
+func containsPort(ports []*ProtocolPort, port *ProtocolPort) bool {
+	for _, protocolPort := range ports {
+		protocol1 := ""
+		switch protocolPort.protocol {
+		case v1beta1.HTTPProtocolType, v1beta1.HTTPSProtocolType, v1beta1.TLSProtocolType, v1beta1.TCPProtocolType:
+			protocol1 = "tcp"
+		default:
+			protocol1 = "udp"
+		}
+		protocol2 := ""
+		switch port.protocol {
+		case v1beta1.HTTPProtocolType, v1beta1.HTTPSProtocolType, v1beta1.TLSProtocolType, v1beta1.TCPProtocolType:
+			protocol2 = "tcp"
+		default:
+			protocol2 = "udp"
+		}
+		if protocol1 == protocol2 && protocolPort.port == port.port {
```

Comment:
nit: you could add
```
if protocolPort.port != port.port {
  continue
}
```
to the top of the loop to skip early

Reply:
A small functional layer4Protocol is extracted to reduce codes, the port number is moved to first to skip early.